app/frontend/src/features/search/useSearchFilters.ts

```diff
@@ -0,0 +1,94 @@
+import { Location } from "history";
+import { useCallback, useEffect, useRef, useState } from "react";
+import { useHistory, useLocation } from "react-router-dom";
+import { UserSearchFilters as ServiceUserSearchFilters } from "service/search";
+
+export interface SearchFilters extends ServiceUserSearchFilters {
+  location?: string;
+}
+
+export default function useSearchFilters(route: string) {
+  const location = useLocation();
+  const [active, setActive] = useState<SearchFilters>(() =>
+    locationToFilters(location)
+  );
+  const pending = useRef<SearchFilters>(active);
+
+  const history = useHistory();
+
+  useEffect(() => {
+    const entries: [string, string][] = [];
+    Object.entries(active).forEach((entry) => {
+      if (typeof entry[1] === "object") {
+        Object.values(entry[1]).forEach((value) => {
+          entries.push([entry[0], `${value}`]);
+        });
+      } else {
+        entries.push(entry);
+      }
+    });
+    const searchParams = new URLSearchParams(entries);
+    history.push(`${route}?${searchParams.toString()}`);
+  }, [active, history, route]);
+
+  const change = useCallback(
+    <T extends keyof SearchFilters>(
+      filter: T,
+      value: Exclude<SearchFilters[T], undefined>
+    ) => {
+      pending.current = { ...pending.current, [filter]: value };
+    },
+    []
+  );
+
+  const remove = useCallback((filter: keyof SearchFilters) => {
+    delete pending.current[filter];
+  }, []);
+
+  const apply = useCallback(() => {
+    setActive(pending.current);
+  }, []);
+
+  return { active, change, remove, apply };
+}
+
+export function locationToFilters(location: Location) {
+  const searchParams = new URLSearchParams(location.search);
+  const filters: SearchFilters = {};
+  Array.from(searchParams.keys()).forEach((untypedkey) => {
+    const key = untypedkey as keyof SearchFilters;
```

Comment:
```suggestion
  Array.from(searchParams.keys()).forEach((key) => {
```
Since you have exhaustively handled all cases anyway with the fallback to a console warn if the type assertion there turns out to be false, is there any point in doing it anyway?

Reply:
I do want to keep it for the autocomplete, but users can also type anything in the url so I feel I should handle default cases.

Reply:
Like you said, since the user can type anything, then `untypedkey`/`key` having the `string` type is correct and so it'd be wrong to type assert? It's not needed for the assignment in the `case`s later since the cases themselves seems to correctly narrow the type of `key` to be just that string, which matches `keyof SearchFilters` and so no errors even without the type assert.